File: 00129. Sum Root to Leaf Numbers/00129.c
// https://leetcode.com/problems/sum-root-to-leaf-numbers

void helper(struct TreeNode *root, int sum, int *ans) {
    if (root == NULL) {
        return;
    }

    sum = sum * 10 + root->val;
    if (root->left == NULL && root->right == NULL) {
        *ans += sum;
    } else {
        helper(root->left, sum,  ans);
        helper(root->right, sum, ans);
    }
}

int sumNumbers(struct TreeNode *root) {
    int ans = 0;

    helper(root, 0, &ans);

    return ans;
}

File: LeetCode/00129. Sum Root to Leaf Numbers/00129.go
// https://leetcode.com/problems/sum-root-to-leaf-numbers

func helper(root *TreeNode, prefix int) int {
    if root == nil {
        return 0
    }

    prefix = prefix * 10 + root.Val

    if root.Left == nil && root.Right == nil {
        return prefix
    }
    return helper(root.Left, prefix) + helper(root.Right, prefix)
}

func sumNumbers(root *TreeNode) int {
    return helper(root, 0)
}
